File: messages/src/main/java/com/epam/deltix/timebase/messages/InstrumentMessage.java
package com.epam.deltix.timebase.messages;

import com.epam.deltix.util.lang.Util;
import java.lang.reflect.InvocationTargetException;

/**
 * Base class for all messages that could be written in Timebase.
 */
public class InstrumentMessage
    implements TimeStampedMessage, MessageInterface, IdentityKey, IdentityKeyInterface, Comparable<InstrumentMessage> {

    /**
     * Time in this field is measured in milliseconds that passed since January 1, 1970 UTC.\
     * For inbound messages special constant {TIMESTAMP_UNKNOWN} marks 'unknown' timestamp in which case TimeBase server stores message using current server time.
     * See also {@link #nanoTime}.
     */
    protected long timestamp = TIMESTAMP_UNKNOWN;

    /**
     * Nanoseconds component of timestamp. Generally speaking value must be in range from [ 0 to 999999 ],
     * but system handles any value (nanoTime simply added to timestamp converted to nanoseconds).
     * Use {@link TimeStamp#getNanoTime(long, int)} helper method calculate message timestamps with nanosecond resolution in the same manner as TimeBase.
     *
     * Note this field is ignored when {@link #timestamp} is set to {TIMESTAMP_UNKNOWN}.
     */
    protected int nanoTime = 0;

    /**
     * Gets message time in milliseconds that passed since January 1, 1970, 00:00:00 GMT
     * @return the number of milliseconds since January 1, 1970, 00:00:00 GMT represented by this message.
     */
    public long getTimeStampMs() {
        return timestamp;
    }

    /**
     * Sets this message to represent a point in time that is
     * <code>value</code> milliseconds after January 1, 1970 00:00:00 GMT.
     *
     * For inbound messages special constant {link TIMESTAMP_UNKNOWN} marks 'unknown' timestamp
     * in which case TimeBase server stores message using current server time.
     *
     * @param   value   the number of milliseconds.
     */

    public void setTimeStampMs(long value) {
        this.timestamp = value;
    }

    /**
     * Indicated that this message has defined point in time.
     */
    public boolean hasTimeStampMs() {
        return timestamp != TIMESTAMP_UNKNOWN;
    }

    /**
     * Sets this message time as 'undefined' {link TIMESTAMP_UNKNOWN}
     */
    public void nullifyTimeStampMs() {
        this.timestamp = TIMESTAMP_UNKNOWN;
    }

    /**
     * Gets message time in nanoseconds that passed since January 1, 1970, 00:00:00 GMT
     *
     * @return the number of nanoseconds since January 1, 1970, 00:00:00 GMT represented by this message.
     */
    @Override
    public long                 getNanoTime() {
        return TimeStamp.getNanoTime(timestamp, nanoTime);
    }

    /**
     * Indicated that this message has defined nanoseconds component of the time.
     */
    public boolean hasNanoTime() {
        return nanoTime != 0;
    }

    /**
     * Sets this message time as 'undefined' {link TIMESTAMP_UNKNOWN}
     */
    public void nullifyNanoTime() {
        this.nanoTime = 0;
        this.timestamp = TIMESTAMP_UNKNOWN;
    }

    /**
     * Sets this message to represent a point in time that is
     * <code>value</code> nanoseconds after January 1, 1970 00:00:00 GMT.
     *
     * For inbound messages special constant {link TIMESTAMP_UNKNOWN} marks 'unknown' timestamp
     * in which case TimeBase server stores message using current server time.
     *
     * @param   value   the number of nanoseconds.
     */
    public void                 setNanoTime(long value) {
        if (value != TIMESTAMP_UNKNOWN) {
            nanoTime = (int) (value % TimeStamp.NANOS_PER_MS);
            timestamp = value / TimeStamp.NANOS_PER_MS;
        } else {
            timestamp = TIMESTAMP_UNKNOWN;
            nanoTime = 0;
        }
    }

    public int                  compareTime(InstrumentMessage time) {
        return timestamp == time.timestamp ?
                Util.compare(nanoTime, time.nanoTime) : (timestamp > time.timestamp ? 1 : -1);
    }

    public String               getTimeString() {
        return formatNanos(timestamp, (int)nanoTime);
    }

    public static String                formatNanos (long milliseconds, int nanos) {
        if (milliseconds == Long.MIN_VALUE)
            return ("<null>");

        return TicksFormat.format(milliseconds, nanos);
    }

    /**
     * Deep copies content from src instance to this.
     * @param template source for copy.
     */
    @Override
    public InstrumentMessage copyFrom(RecordInfo template) {
        if (template instanceof MessageInfo) {
            MessageInfo t = (MessageInfo) template;
            setTimeStampMs(t.getTimeStampMs());
            setNanoTime(t.getNanoTime());

            // set symbol should copy value
            if (symbol instanceof StringBuilder)
                ((StringBuilder)symbol).setLength(0);
            else
                setSymbol(new StringBuilder());
            ((StringBuilder)symbol).append(t.getSymbol());
        }
        return this;
    }

    /**
     * Creates new instance of this class.
     */
    protected InstrumentMessage createInstance() {
        try {
            return Util.newInstance(getClass());
        } catch (ClassNotFoundException | NoSuchMethodException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
            throw new RuntimeException("Cannot create instance for the " + getClass());
        }
    }

    /**
     * Creates copy of this instance.
     * @return copy.
     */
    @Override
    public InstrumentMessage clone() {
        InstrumentMessage c = createInstance();
        c.copyFrom(this);
        return c;
    }

    protected CharSequence         symbol = "";

    /**
     * Instrument name.
     * @return String representation of identity name.
     */
    public final CharSequence         getSymbol () {
        return symbol;
    }

    /**
     * Sets Instrument name.
     * @param value - Symbol
     */
    public final void setSymbol(CharSequence value) {
        this.symbol = value;
    }

    /**
     * Returns true, if instrument name is defined.
     * @return true if name is not null
     */
    public final boolean hasSymbol() {
      return symbol != null;
    }

    /**
     * Set instrument name to 'undefined' value
     */
    public final void nullifySymbol() {
      this.symbol = null;
    }

//    /**
//     * Method copies state of given template into this object
//     * @param deep if true performs deep copy of mutable properties
//     */
//    @Deprecated
//    public void                 copy (Object template, boolean deep) {
//
//        if (template instanceof InstrumentMessage) {
//            InstrumentMessage t = (InstrumentMessage) template;
//            timestamp = t.timestamp;
//            nanoTime = t.nanoTime;
//        }
//
//        if (template instanceof InstrumentIdentity) {
//            InstrumentIdentity   sm = (InstrumentIdentity) template;
//
//            instrumentType = sm.getInstrumentType ();
//            symbol = sm.getSymbol ();
//            if (deep && symbol != null)
//                symbol = symbol.toString ();
//        }
//    }

    public int                  compareTime(TimeStamp time) {
        // Do not change. JIT compile this implementation into highly optimized branch free code with 4 cmovnl instructions.
        // If nanos does not belong to 0..999999 range or timestamp does not belong to 0..2^53
        // then timestamp comparison result is Undefined
        long nanos1 = timestamp * TimeStamp.NANOS_PER_MS + nanoTime;
        long nanos2 = time.timestamp * TimeStamp.NANOS_PER_MS + time.nanosComponent;
        nanos1 -= nanos2;
        return (nanos1 > 0 ? 1 : 0) - (nanos1 < 0 ? 1 : 0);
    }
    
    @Override
    public int                  compareTo(InstrumentMessage o) {
        if (timestamp == o.timestamp) {
            if (nanoTime == o.nanoTime)
                return Util.fastCompare(symbol, o.symbol);
            return nanoTime > o.nanoTime ? 1 : -1;
        }

        return timestamp > o.timestamp ? 1 : -1;
    }

    public String               toString () {
        StringBuilder sb = new StringBuilder();
        return toString(sb).toString();
    }

    public StringBuilder        toString(StringBuilder str) {
        str.append("{ \"$type\": \"InstrumentMessage\"");

        if (hasSymbol())
            str.append(", \"symbol\": \"").append(symbol).append("\"");

        if (hasTimeStampMs())
            str.append(", \"timestamp\": \"").append(getTimeString()).append("\"");

        str.append("}");

        return str;
    }


    /**
     * Indicates whether some other object is "equal to" this one.
     */
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof MessageInfo))
            return false;
        MessageInfo other = (MessageInfo)obj;
        if (getNanoTime() != other.getNanoTime())
            return false;
        if ((symbol == null) != (other.getSymbol() == null))
            return false;
        if (symbol != null && !CharSequenceUtils.equals(symbol, other.getSymbol()))
            return false;

        return true;
    }


    /**
     * Reset all instance field to their default states.
     * @return this.
     */
    @Override
    public InstrumentMessage reset() {
        symbol = "";
        timestamp = TIMESTAMP_UNKNOWN;
        nanoTime = 0;
        return this;
    }

    /**
     * Set null to all fields of this instance.
     * @return this.
     */
    @Override
    public InstrumentMessage nullify() {
        nullifySymbol();
        nullifyTimeStampMs();
        nullifyNanoTime();
        return this;
    }
}

File: messages/csharp/src/Deltix.Timebase.Messages/EPAM.Deltix.Timebase.Messages/InstrumentMessage.cs
/*
  Copyright 2021 EPAM Systems, Inc

  See the NOTICE file distributed with this work for additional information
  regarding copyright ownership. Licensed under the Apache License,
  Version 2.0 (the "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
  License for the specific language governing permissions and limitations under
  the License.
 */
using com.epam.deltix.containers.;
using System;
using System.Collections.Generic;
using System.Text;

namespace EPAM.Deltix.Timebase.Messages{
	/// <summary>
	/// Base class for all messages that could be written in Timebase.
	/// </summary>
	[Serializable]
	[OldElementName(Value = "deltix.qsrv.hf.pub.InstrumentMessage")]
	[SchemaElement(Name = "deltix.timebase.messages.InstrumentMessage", Title = "Instrument Message", Description = null)]
	public class InstrumentMessage : IMessageInterface, IMessageInfo, IRecordInfo, IRecordInterface, IIdentityKey, IIdentityKeyInterface, ICloneable
	{
		public static string ClassName = "deltix.timebase.messages.InstrumentMessage";

		private HdDateTime _timestamp = TypeConstants.TimestampNull;		

		private string _symbol = "";

		/// <summary>
		/// Time in this field is measured in nanoseconds that passed since January 1, 1970 UTC.
		/// For inbound messages special constant {link TIMESTAMP_UNKNOWN} marks 'unknown' timestamp in which case TimeBase server stores message using current server time.
		/// </summary>
		public virtual HdDateTime NanoTime
		{
			get
			{
				return this._timestamp;
			}
			set
			{
				this._timestamp = value;
			}
		}

		/// <summary>
		/// Sets null to Timestamp property.
		/// </summary>
		public void NullifyNanoTime()
		{
			this._timestamp = TypeConstants.TimestampNull;
		}

		public bool HasNanoTime()
		{
			return this._timestamp != (HdDateTime)TypeConstants.TimestampNull;
		}

		/// <summary>
		/// Time in this field is measured in milliseconds that passed since January 1, 1970 UTC.
		/// For inbound messages special constant {link TIMESTAMP_UNKNOWN} marks 'unknown' timestamp in which case TimeBase server stores message using current server time.
		/// </summary>

		public DateTime TimeStampMs
		{
			get { return _timestamp.Timestamp;  }
			set { _timestamp = new HdDateTime(value); }
		}

		/// <summary>
		/// Instrument name.
		/// </summary>
		public string Symbol
		{
			get
			{
				return this._symbol;
			}
			set
			{
				this._symbol = value;
			}
		}

		/// <summary>
		/// Flag that indicates whether property Timestamp has an assigned value.
		/// <returns>true if property has an assigned value, false otherwise.</returns>
		/// </summary>
		public bool HasTimeStampMs()
		{
			return this._timestamp != (HdDateTime)TypeConstants.TimestampNull;
		}

		/// <summary>
		/// Sets null to Timestamp property.
		/// </summary>
		public void NullifyTimeStampMs()
		{
			this._timestamp = TypeConstants.TimestampNull;
		}

		/// <summary>
		/// Flag that indicates whether property Symbol has an assigned value.
		/// <returns>true if property has an assigned value, false otherwise.</returns>
		/// </summary>
		public bool HasSymbol()
		{
			return this._symbol != null;
		}

		/// <summary>
		/// Sets null to Symbol property.
		/// </summary>
		public void NullifySymbol()
		{
			this._symbol = null;
		}

		/// <summary>
		/// Creates a new instance of message.
		/// <returns>New instance of message of the same type as this.</returns>
		/// </summary>
		protected virtual IRecordInterface CreateInstance()
		{
			return new InstrumentMessage();
		}

		/// <summary>
		/// Creates a new instance of message.
		/// <returns>New instance of message of the same type as this.</returns>
		/// </summary>
		protected virtual IRecordInfo CloneImpl()
		{
			IRecordInterface message = this.CreateInstance();
			message.CopyFrom(this);
			return message;
		}

		/// <summary>
		/// Creates copy of this instance.
		/// <returns>Copy.</returns>
		/// </summary>
		object ICloneable.Clone()
		{
			return this.CloneImpl();
		}

		/// <summary>
		/// Creates copy of this instance.
		/// <returns>Copy.</returns>
		/// </summary>
		IRecordInterface IRecordInterface.Clone()
		{
			return (IRecordInterface)this.CloneImpl();
		}

		/// <summary>
		/// Creates copy of this instance.
		/// <returns>Copy.</returns>
		/// </summary>
		IRecordInfo IRecordInfo.Clone()
		{
			return this.CloneImpl();
		}

		/// <summary>
		/// Creates copy of this instance.
		/// <returns>Copy.</returns>
		/// </summary>
		IMessageInfo IMessageInfo.Clone()
		{
			return (IMessageInfo)this.CloneImpl();
		}

		/// <summary>
		/// Creates copy of this instance.
		/// <returns>Copy.</returns>
		/// </summary>
		IMessageInterface IMessageInterface.Clone()
		{
			return (IMessageInterface)this.CloneImpl();
		}

		/// <summary>
		/// Creates copy of this instance.
		/// <returns>Copy.</returns>
		/// </summary>
		public InstrumentMessage Clone()
		{
			return (InstrumentMessage)this.CloneImpl();
		}

		/// <summary>
		/// Deep copies content from src instance to this.
		/// <param name="source">Source for copy.</param>
		/// <returns>this.</returns>
		/// </summary>
		protected virtual IRecordInterface CopyFromImpl(IRecordInfo source)
		{
			if (typeof(IMessageInfo).IsInstanceOfType(source))
			{
				IMessageInfo typedSource = (IMessageInfo)source;
				this.NanoTime = typedSource.NanoTime;				
				this.Symbol = typedSource.Symbol;
			}
			return this;
		}

		/// <summary>
		/// Deep copies content from src instance to this.
		/// <param name="source">Source for copy.</param>
		/// <returns>this.</returns>
		/// </summary>
		IRecordInterface IRecordInterface.CopyFrom(IRecordInfo source)
		{
			return this.CopyFromImpl(source);
		}

		/// <summary>
		/// Deep copies content from src instance to this.
		/// <param name="source">Source for copy.</param>
		/// <returns>this.</returns>
		/// </summary>
		IMessageInterface IMessageInterface.CopyFrom(IRecordInfo source)
		{
			return (IMessageInterface)this.CopyFromImpl(source);
		}

		/// <summary>
		/// Deep copies content from src instance to this.
		/// <param name="source">Source for copy.</param>
		/// <returns>this.</returns>
		/// </summary>
		public InstrumentMessage CopyFrom(IRecordInfo source)
		{
			return (InstrumentMessage)this.CopyFromImpl(source);
		}

		/// <summary>
		/// Set null to all fields of this instance.
		/// <returns>this.</returns>
		/// </summary>
		protected virtual IRecordInterface NullifyImpl()
		{
			this.NullifyTimeStampMs();			
			this.NullifySymbol();
			return this;
		}

		/// <summary>
		/// Set null to all fields of this instance.
		/// <returns>this.</returns>
		/// </summary>
		IRecordInterface IRecordInterface.Nullify()
		{
			return this.NullifyImpl();
		}

		/// <summary>
		/// Set null to all fields of this instance.
		/// <returns>this.</returns>
		/// </summary>
		IMessageInterface IMessageInterface.Nullify()
		{
			return (IMessageInterface)this.NullifyImpl();
		}

		/// <summary>
		/// Set null to all fields of this instance.
		/// <returns>this.</returns>
		/// </summary>
		public InstrumentMessage Nullify()
		{
			return (InstrumentMessage)this.NullifyImpl();
		}

		/// <summary>
		/// Reset all instance field to their default states.
		/// <returns>this.</returns>
		/// </summary>
		protected virtual IRecordInterface ResetImpl()
		{
			this._timestamp = TypeConstants.TimestampNull;
			this._symbol = "";
			return this;
		}

		/// <summary>
		/// Reset all instance field to their default states.
		/// <returns>this.</returns>
		/// </summary>
		IRecordInterface IRecordInterface.Reset()
		{
			return this.ResetImpl();
		}

		/// <summary>
		/// Reset all instance field to their default states.
		/// <returns>this.</returns>
		/// </summary>
		IMessageInterface IMessageInterface.Reset()
		{
			return (IMessageInterface)this.ResetImpl();
		}

		/// <summary>
		/// Reset all instance field to their default states.
		/// <returns>this.</returns>
		/// </summary>
		public InstrumentMessage Reset()
		{
			return (InstrumentMessage)this.ResetImpl();
		}

		/// <summary>
		/// Determines whether the specified object is equal to the current object.
		/// <param name="obj">The object to compare with the current object.</param>
		/// <returns>true if the specified object is equal to the current object; otherwise, false.</returns>
		/// </summary>
		public virtual bool Equals(InstrumentMessage obj)
		{
			if (obj == null)
			{
				return false;
			}
			if (this.NanoTime != obj.NanoTime)
			{
				return false;
			}			
			if (this.Symbol != obj.Symbol)
			{
				return false;
			}
			return true;
		}

		/// <summary>
		/// Determines whether the specified object is equal to the current object.
		/// <param name="obj">The object to compare with the current object.</param>
		/// <returns>true if the specified object is equal to the current object; otherwise, false.</returns>
		/// </summary>
		public override bool Equals(object obj)
		{
			if (!typeof(InstrumentMessage).IsInstanceOfType(obj))
			{
				return false;
			}
			return this.Equals((InstrumentMessage)obj);
		}

		/// <summary>
		/// Calculates a hash code for the current object.
		/// <returns>A hash code for the current object.</returns>
		/// </summary>
		public override int GetHashCode()
		{
			int hash2 = -2128831035;
			hash2 = hash2 * 16777619 + this.NanoTime.GetHashCode();
			return hash2 * 16777619 + this.Symbol.GetHashCode();
		}

		/// <summary>
		/// Appends an object state to a given StringBuilder in a form of JSON.
		/// <returns>A StringBuilder that was used to append the object to.</returns>
		/// </summary>
		public virtual StringBuilder ToString(StringBuilder builder)
		{
			builder.Append("{ \"$type\":  \"InstrumentMessage\"");
			if (this.HasTimeStampMs())
			{
				builder.Append(", \"Timestamp\": \"");
				builder.Append(this.NanoTime);
				builder.Append("\"");
			}			
			if (this.HasSymbol())
			{
				builder.Append(", \"Symbol\": \"");
				builder.Append(this.Symbol);
				builder.Append("\"");
			}
			builder.Append("}");
			return builder;
		}

		/// <summary>
		/// Returns a string that represents the current object.
		/// <returns>A string that represents the current object.</returns>
		/// </summary>
		public override string ToString()
		{
			return this.ToString(new StringBuilder()).ToString();
		}
	}
}